### PR TITLE
[Core][EPLB] Refactor step to unify synchronous and asynchronous expert rearrangement

### PR DESCRIPTION
**Pull request by 845473182**
# Summary

This PR introduces an **asynchronous Expert Parallel Load Balancer (EPLB)** mechanism along with a background process (`EplbProcess`) and expert mapping module (`EplbExpertMapper`).  
It refactors `EplbState` to support non-blocking expert rearrangement and introduces new modules for expert transfer planning and asynchronous communication.  
The asynchronous framework allows overlap between load balancing and computation, improving scalability and reducing synchronization overhead.

## Purpose

- Extend the existing EPLB implementation with **asynchronous processing** to overlap expert redistribution with forward computation.
- Add `EplbProcess` to manage background multiprocessing tasks for expert rearrangement and weight transfer.
- Add `EplbExpertMapper` providing modular migration strategies (`BipartiteExpertUpdate`, `GreedyExpertUpdate`) to plan expert send/receive mappings across ranks.
- Refactor `EplbState` to integrate asynchronous hooks and buffer preallocation for non-blocking D2D communication.
- Improve modularity and maintainability of the EPLB system while keeping full backward compatibility (`enable_async=False` → original behavior).

### Added `eplb_process.py`
Implements the asynchronous process controller for expert rearrangement.
- Handles task lifecycle: initialization, submission, execution, and cleanup.
- Uses multiprocessing with inter-process queues for argument/result exchange.
- Provides result packaging with expert send/receive info and updated expert maps.
- Isolates subprocess errors and reports them back to the parent process safely.

### Added `eplb_expert_mapper.py`
Introduces modular expert mapping strategies for flexibility and efficiency.
- **`BipartiteExpertUpdate`** – minimizes redundant transfers via bipartite matching.
- **`GreedyExpertUpdate`** – lightweight greedy strategy for faster heuristic mapping.
- Each generates per-layer expert migration plans (send/recv dicts and updated maps).

### Updated `eplb_state.py`
Refactored to integrate asynchronous control flow.
- Added async hooks: `step_before_forward`, `shuffle_layer_async`, `generate_expert_d2d_transfer_task`.
- Introduced `buffer_tensor_list` for preallocating expert weight buffers.
- Integrated coordination between load metrics collection, async submission, and expert map update.
- Added `enable_async` config flag in EPLB state for optional background execution.
- Ensured backward compatibility with original synchronous EPLB.

### eplb_process.py
Implements asynchronous EPLB process control.  
Manages background expert rearrangement, inter-process communication, and result synchronization.

### eplb_expert_mapper.py
Implements expert migration strategies: `BipartiteExpertUpdate` and `GreedyExpertUpdate`.  
Each strategy generates layer-wise expert movement plans based on global load metrics.

### eplb_state.py
Refactored core EPLB state manager to support asynchronous mode.  
Coordinates buffer initialization, task scheduling, and map updates across ranks.

## Test Plans

baseline:
```
python -m vllm.entrypoints.openai.api_server \
  --model $MODEL \
  --tensor-parallel-size 2 --data-parallel-size 2 \
  --enable-expert-parallel --enable-eplb \
  --eplb-config '{"window_size":100,"step_interval":100,”enable_async":false}' \
  --max-num-seqs 512
```

this pr:
```
python -m vllm.entrypoints.openai.api_server \
  --model $MODEL \
  --tensor-parallel-size 2 --data-parallel-size 2 \
  --enable-expert-parallel --enable-eplb \
  --eplb-config '{"window_size":100,"step_interval":100,"num_wait_worker_iterations":500,"enable_async":true}' \
  --max-num-seqs 512
```

dataset: gsm8k

model: Qwen3-30B-A3B-Instruct-2507-FP8

tool: ais_bench

benchmark_config:
```
models = [
    dict(
        attr="service",
        type=VLLMCustomAPIChatStream,
        abbr='vllm-api-stream-chat',
        path="/path/to/model",
        model="qwen3",
        request_rate=1,
        retry=2,
        host_ip="localhost",
        host_port=8008,
        max_out_len=4096,
        batch_size=32
        trust_remote_code=False,
        generation_kwargs=dict(
            temperature=0.5,
            top_k=10,
            top_p=0.95,
            seed=None,
            repetition_penalty=1.03,
            ignore_eos=False,
        ),
        pred_postprocessor=dict(type=extract_non_reasoning_content)
    )
]
```

## Test Results

baseline:
baseline

this pr:
async

### CODE REVIEW

File: vllm/config/parallel.py [hunk 1]
```diff
@@ -61,6 +61,23 @@ class EPLBConfig:
...
+    A gate to trigger asynchronous expert load rebalancing and weight transfer
+    """
+
+    expert_mapper_policy_type: int = 0
+    """
+    Determine using which expert mapping strategy
+
+    0: GreedyExpertUpdate
+    1: BipartiteExpertUpdate
+    """
```
Comment (hmellor) [inline vllm/config/parallel.py]: Magic numbers.

Can we use

```py
ExpertMapperPolicy = Literal["greedy", "bipartite"]
...
    expert_mapper_policy_type: ExpertMapperPolicy = "greedy"
```
instead?

File: vllm/distributed/eplb/eplb_expert_mapper.py [hunk 2]
```diff
@@ -1,13 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+import networkx as nx
 import numpy as np
 import torch
-import networkx as nx
+
+
+from typing import Dict, List
```
Comment (hmellor) [inline vllm/distributed/eplb/eplb_expert_mapper.py]: Deprecated typing that will cause `pre-commit` to fail. Please just directly hint with `dict` and `list`

File: vllm/distributed/eplb/eplb_state.py [hunk 3]
```diff
@@ -51,6 +51,19 @@
...
+    global_expert_load_window: torch.Tensor
+    num_replicas: int
+    num_groups: int
+    num_nodes: int
+    num_gpus: int
+
+@dataclass
+class ExpertMapperArgs:
+    num_moe_layers: int
+    policy_type: Literal["greedy","bipartite"]
```
Comment (hmellor) [inline vllm/distributed/eplb/eplb_state.py]: So that this doesn't need to be duplicated here, you can define 

```py
ExpertMapperPolicy = Literal["greedy", "bipartite"]
```

in `config/parallel.py` and import it here

File: vllm/config/parallel.py [hunk 4]
```diff
@@ -71,14 +71,15 @@ class EPLBConfig:
     A gate to trigger asynchronous expert load rebalancing and weight transfer
     """
 
-    expert_mapper_policy_type: int = 0
+    expert_mapper_policy_type: Literal["greedy", "bipartite"] = "greedy"
```
Comment (hmellor) [inline vllm/config/parallel.py]: `_type` doesn't add anything and will make all lines longer

```suggestion
    expert_mapper_policy: Literal["greedy", "bipartite"] = "greedy"
```

**Review by hmellor — CHANGES_REQUESTED** [review]
